streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -690,6 +693,7 @@ public StreamThread(final Time time,
         this.log = logContext.logger(StreamThread.class);
         this.rebalanceListener = new RebalanceListener(time, taskManager, this, this.log);
         this.taskManager = taskManager;
+        this.producer = producer;
```

Comment:
Hmm... note that when EOS is turned on, each task will has its own producer client and the producer object passed in here will be null.

So I'd suggest update the following code in line 1224: when `producer == null`, try to iterate the owned tasks from the `taskManager.activeTasks().values()` and get its producer (it is private, so we may need to add a package private getter).

In this way for both EOS and non-EOS we will get the producer metrics.

Reply:
Make sense to me!